src/key-value-engine/B-Tree_vs_The_Log-Structured_Merge-Tree.md

```diff
@@ -0,0 +1,101 @@
+
+For example, if you need to read 5 pages to answer a query, read amplification is 5.
+
+### Space Amplification
+
+`Space amplification` is the ratio of the amount of data on storage medium compared to the amount of data on database. 
+
+For example, if you Put 10MB in the database and it uses 100MB on disk, then the space amplification is 10.
+
+## Analysis
+
+The term B-tree may refers to a specific design or it may refers to a general class of designs. In the narrow sense, a B-tree stores keys in its internal nodes but need not store those keys in the records at the leaves. The [B+ tree](https://en.wikipedia.org/wiki/B%2B_tree#Insertion) is one of the most famous variation of B-tree, behind which the idea is that internal nodes only contains keys, and to which an additional level which contains values is added at the bottom with linked leaves.
+
+LSM-tree performs `compaction`  to merges several sstables into one new sstable which contains only the live data from the input sstables. Compation helps LSM-tree to recycle space and reduce read amplification.  There are two kinds of `compaction strategy` which is `Size-Tiered compaction strategy` (STCS) and `Level-Based compaction strategy` (LBCS). The idea behind STCS is compacting small sstables into medium sstables when LSM-tree has enough small sstables and compacting medium sstables into large sstables when LSM-tree has enough medium sstables. The idea of LBCS is to organize data into levels which contains one sorted run.
+
+This section is going to discuss about the write amplification and read amplification of B+tree and Level-Based LSM-tree. 
+
+### B+ Tree
+
+In the B+ tree, copies of the keys are stored in the internal nodes; the keys and records are stored in leaves; in addition, a leaf node may include a pointer to the next leaf node to speed sequential access.
+
+To simplify the analysis, let's say that the block size of the tree as `B`, and assumes that keys, pointers, and records are constant sized, so that each internal node contains `O(B)` children and each leaf contains `O(B)` data records. (The root node is a special case, and can be nearly empty in some situations.) Under all these assumptions, the depth of a B tree is 
+$$
+O(log_BN/B)
+$$
+where `N` is the size of database.
+
+#### Write Amplification
+
+For the worst-case insertion workloads, every insertion requires writing the leaf block containing the record, so the write amplification is `B`.
+
+#### Read Amplification
+
+The number of I/O's for any query is at most `O(logB N/B)` which is the depth of the tree. 
+
+### Level-Based LSM-tree
+
+In level-based LSM-tree, data is organized into levels. Each level contains one run. Data starts in level 0, then gets merged into the level 1 run. Eventually the level 1 run is merged into the level 2 run, and so forth. Each level is constrained in its sizes. Growth factor `k`  is specified as the magnification of data size at each level 
+$$
```

Comment:
what is '$'?

Reply:
It's a mark for LaTex. FYI: https://github.com/cben/mathdown/wiki/math-in-markdown